### ISSUE TITLE
Please update the PHP environment

### ISSUE THREAD

**Issue by stof** [travis-ci/travis-ci#4897]
You announced a while ago that the plan was to update the environment every 2 months. But as far as I can tell, the PHP environment has not been updated since February (Composer reports ``2015-02-11`` as its date when showing its version at the beginning of builds).

When is an update planned ?

**Comment by joshk** [comment]
Thanks for opening this @stof 

We are in the process of reworking our build env update process, which hit some snags after out last update.

We hope to have some more info to share soon.

**Comment by cybernet** [comment]
+1

**Comment by acoulton** [comment]
Any further news? Composer is still showing `2015-02-11` as the last updated version.

**Comment by theofidry** [comment]
Any update on this issue? Composer doesn't seem to be updated yet.

**Comment by mizunashi-mana** [comment]
My composer and travis config can run on PHP 7.0/nightly, but cannot run on 5.6/HHVM.
The reason is 5.6's composer is old version's...

Latest composer includes https://github.com/composer/composer/issues/4852 and some fixes.

Please, update!

**Comment by BanzaiMan** [comment]
@mizunashi-mana The pre-installed versions (5.6, for example) will not receive updates until the entire build environment is updated.

Until then, either run `composer self-update` yourself, or specify TEENY version (e.g., `5.6.19`).

**Comment by stof** [comment]
@BanzaiMan according to the date of the Composer build, this means that the last build environment update was in February 2015. You announced updates every 2 months last year. What is the actual update plans ?

**Comment by theofidry** [comment]
It is also explicitly [in the documentation](https://docs.travis-ci.com/user/languages/php#Installing-Composer-packages) that it is updated every 30-60 days.

**Comment by hkdobrev** [comment]
@joshk @BanzaiMan @theofidry More than 60 days passed since the last update of Composer on Travis CI (it was around March 4th). During that time Composer maintainers launched 1.0.0 and 1.1.0, introduced update channels and @Seldaek suggests using `composer self-update --preview` in CI environments. Since the current Composer version on Travis CI is pre-1.0 it does not support the update channels and the new default update logic to only stable versions on Composer. So doing `composer self-update` (which seems to be done automatically by Travis CI now) updates to the latest snapshot dev version of Composer instead of the latest tagged version. Could you please update the default version of Composer in Travis CI images and use `composer self-update --preview` by default? This would be quite future-proof even if you don't update Composer every 60 days in the image. This would probably need adjusting for Composer 2.0.

Thank you!

**Comment by theofidry** [comment]
`composer self-update --preview` might not be ideal. Personally I'm quite satisfied having `composer self-update` automatically now to be honest. To me it's more future proof as `--preview` because once stable, it will keep to the stable version, not a pre-release. If you absolutely need the pre-release, you should update it manually.

So IMO this ticket could closed :)

**Comment by hkdobrev** [comment]
@theofidry I will accept `composer self-update` as well, keeping it to the stable channel, but still Composer needs to be updated in the image first before this issue is closed. At the moment Composer on Travis CI is pre-1.0 and doesn't know about channels and `composer self-update` is updating to the latest snapshot.

**Comment by hkdobrev** [comment]
@joshk @BanzaiMan Hi, sorry for spamming, but the default Composer version on Travis CI is still pre-1.0.0 so it has no idea of update channels and Travis CI is automatically updating it to the latest snapshot. I don't see a reason not to update Composer as you've pledged before you'll do every 60 days.

**Comment by theofidry** [comment]
@hkdobrev composer is self update in every build: https://travis-ci.org/hautelook/AliceBundle/jobs/136010841#L167

**Comment by hkdobrev** [comment]
@theofidry The version which is installed initially is the one which is doing the self updating. The self-updater pre-1.0.0 doesn't know about channels. So it always _upgrades_ Composer to the latest commit on master. If there is a breaking change in Composer, Travis CI would take it and self-update to it.

**Comment by stof** [comment]
which means that to get a stable build, you need to perform a second self-update from the post-1.0 version installed by the first one. Some projects are doing it. Not very friendly though...

**Comment by hkdobrev** [comment]
@stof This doesn't help as you are already on the latest version. Not sure if `composer self-update --preview` would downgrade.

**Comment by stof** [comment]
@hkdobrev no. ``composer self-update`` will update to the latest stable, as the pre-1.0 build will not create a file remembering that it updated to the snapshot channel) and ``--stable`` is the default when no option is provided and nothing is remembered.

**Comment by theofidry** [comment]
@stof in the build I've linked you can see that the version used is `1.2-dev` after `composer self-update`.

**Comment by stof** [comment]
@theofidry yeah, please re-read my comment. The workaround is to run self-update *a second time*, which is not done in the build you linked

**Comment by theofidry** [comment]
> The workaround is to run self-update a second time, which is not done in the build you linked

Yeah, which brings back to the original problem: having a composer update to date to avoid all users to have to manually do it themselves. Looks even more inefficient now because you now 2 self-update done.

**Comment by BanzaiMan** [comment]
Sorry for the delay in updating the build environments. We are carrying out lower-level tasks to make the future updates easier. At this time, we cannot provide you with ETA for the build environment updates, or, once we get on track, how often the updates will occur.

I understand that this can be a problem for PHP builds. Is it agreeable to run `composer self-update` twice in the build in the meantime?

**Comment by stof** [comment]
@BanzaiMan I think documenting the workaround only may be enough, so that people wanting to use a stable version know about it. People were using unstable versions of Composer for years on Travis, and so things should be fine for most of them with a single update rather than 2. Making the double self-update for everyone may not be necessary.

The best solution would still be to have an environment update of course (more than 1-year old is very old). Too bad that composer cannot be updated in environments themselves (at least, we are getting newer PHP versions now thanks to on-demand installation)

**Comment by hkdobrev** [comment]
> The best solution would still be to have an environment update of course (more than 1-year old is very old).

Just for reference, the Composer version installed initially in the VM is from March 4th 2016.

So the PHP build environment and the Composer version are in some way decoupled.

**Comment by BanzaiMan** [comment]
Composer version will depend on the build image provisioning time.

**Comment by hkdobrev** [comment]
Composer version is now 1.1.1 by default on Travis CI builds with PHP 7! However PHP 5.6 is still using an old version.

**Comment by alexislefebvre** [comment]
Composer on PHP 5.6 uses a stable release (1.2.3 [10 days ago](https://travis-ci.org/alexislefebvre/AsyncTweetsBundle/jobs/181197681#L150)).

But Composer on PHP 7 uses a unstable release (1.3-dev [10 days ago](https://travis-ci.org/alexislefebvre/AsyncTweetsBundle/jobs/181197684#L169)). I don't want Travis CI to use a `dev` version. [Installing Composer from scratch](https://github.com/alexislefebvre/AsyncTweetsBundle/pull/28/files#diff-354f30a63fb0907d4ad57269548329e3) circumvents the problem but I think that it would be better if Travis CI always provide a stable release of Composer.

**Comment by akinofftz** [comment]
Currently, all Travis builds of Composer projects are broken, because Composer's `master` is broken (see composer/composer#6021). So I would recommend switching back to stable releases.

**Comment by stof** [comment]
@akinofftz your comment is far too alarming. This issue affects only projects using inline package repositories. So it is far from being all projects. Thus, it does not affect installations from a lock file either.
